### PR TITLE
Job may get marked as 'Never run'

### PR DESCRIPTION
**Pull request by arungrover**
#### Bug/feature Description
In some cases when a job exercises node buckets code, it may get marked as 'Can Never run' even when the job could run on the resources available in the complex.

#### Affected Platform(s)
All

#### Cause / Analysis / Design
In node buckets code when a bucket is looked and it is found that resources requested by the job does not exist in the bucket, the error status is marked as 'NEVER_RUN'. This error structure is not cleared when scheduler check the next available node bucket. This means if the last bucket that was seen by scheduler results into a 'NEVER_RUN' error then that error never gets cleared up an persists with the job. This eventually marks the job as a never run job.

#### Solution Description
Added code to clear up error structure and also fixed a problem where flags passed to is_ok_to_run was not getting cleared when a non-node-bucket job gets encountered.

#### Testing logs/output
[test_before_fix.txt](https://github.com/PBSPro/pbspro/files/2532064/test_before_fix.txt)
[test_after_fix.txt](https://github.com/PBSPro/pbspro/files/2532065/test_after_fix.txt)

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: test/tests/functional/pbs_node_buckets.py [hunk 1]
```diff
@@ -895,3 +895,45 @@ def test_booleans(self):
         for n in ev:
             self.server.expect(
                 NODE, 'resources_available.bool', op=UNSET, id=n)
+
+    @timeout(450)
+    def test_last_pset_can_never_run(self):
+        """
+        Test that the job does not retain the error value of last bucket
+        seen by the node bucketing code. To check this make sure that the last
+        bucket check results into a 'can never run' case because resources do
```
Comment (bhroam) [inline test/tests/functional/pbs_node_buckets.py]: It isn't the last bucket checked, it is the last placement set checked.

File: test/tests/functional/pbs_node_buckets.py [hunk 2]
```diff
@@ -895,3 +895,45 @@ def test_booleans(self):
...
+        not match and check that the job is not marked as never run.
+        """
+
+        self.server.manager(MGR_CMD_CREATE, RSC,
+                            {'type': 'long', 'flag': 'nh'}, id='foo')
+        self.server.manager(MGR_CMD_CREATE, RSC,
+                            {'type': 'string', 'flag': 'h'}, id='bar')
+        self.server.manager(MGR_CMD_SET, SERVER, {'node_group_key': 'bar'})
+        self.server.manager(MGR_CMD_SET, SERVER, {'node_group_enable': 't'})
```
Comment (bhroam) [inline test/tests/functional/pbs_node_buckets.py]: Spell out True

File: test/tests/functional/pbs_node_buckets.py [hunk 3]
```diff
@@ -895,3 +895,45 @@ def test_booleans(self):
...
+             'resources_available.bar': 'large'}
+        self.server.create_vnodes(name='vnode', attrib=a, num=8,
+                                  mom=self.mom, sharednode=False,
+                                  expect=False)
+        self.scheduler.add_resource('foo')
+        a['resources_available.foo'] = 8
+        a['resources_available.ncpus'] = 8
+        a['resources_available.bar'] = 'small'
+        for val in range(0, 5):
+            vname = "vnode["+str(val)+"]"
```
Comment (bhroam) [inline test/tests/functional/pbs_node_buckets.py]: Put spaces around the pluses

File: test/tests/functional/pbs_node_buckets.py [hunk 4]
```diff
@@ -895,3 +895,45 @@ def test_booleans(self):
...
+            self.server.manager(MGR_CMD_SET, NODE, a, id=vname)
+        chunk1 = '4:ncpus=5:foo=5'
+        a = {'Resource_List.select': chunk1,
+             'Resource_List.place': 'scatter:excl'}
+        j1 = Job(TEST_USER, attrs=a)
+        j1.set_sleep_time(10000)
```
Comment (bhroam) [inline test/tests/functional/pbs_node_buckets.py]: Will the default sleep work here?

File: src/scheduler/buckets.c [hunk 5]
```diff
@@ -1100,16 +1098,28 @@ check_node_buckets(status *policy, server_info *sinfo, queue_info *qinfo, resour
...
 
+			clear_schd_error(err);
 			nspecs = map_buckets(policy, nodepart[i]->bkts, resresv, err);
 			if (nspecs != NULL)
 				return nspecs;
-			if (err->status_code == NOT_RUN)
+			if (err->status_code == NOT_RUN) {
+				if (failerr->status_code == SCHD_UNKWN)
+					move_schd_error(failerr, err);
 				can_run = 1;
```
Comment (agrawalravi90) [inline src/scheduler/buckets.c]: i know it's not your code, but do you know why 'can_run' is being set to 1 if err status code is NOT_RUN ?

File: src/scheduler/buckets.c [hunk 6]
```diff
@@ -1121,7 +1131,8 @@ check_node_buckets(status *policy, server_info *sinfo, queue_info *qinfo, resour
 				schdlog(PBSEVENT_DEBUG3, PBS_EVENTCLASS_JOB, LOG_DEBUG, resresv->name,  "Request won't fit into any placement sets, will use all nodes");
 				return map_buckets(policy, sinfo->buckets, resresv, err);
 			}
-		}
+		} else
+			move_schd_error(err, failerr);
```
Comment (agrawalravi90) [inline src/scheduler/buckets.c]: I'm not clear on the purpose of 'can_run', so I apologize for my ignorance, but if the job can run, won't we just return from the function on line 1117 ?

Comment (agrawalravi90) [reply src/scheduler/buckets.c]: regardless, can you please add a comment here to explain why you are moving failerr to err if can_run == 1 ?

File: src/scheduler/buckets.c (hunk 5 above)
Comment (bhroam) [reply src/scheduler/buckets.c]: NOT_RUN means the job can possibly run but not right now.  NEVER_RUN means the job can never run.  The variable is saying the job can possibly run.

File: src/scheduler/buckets.c (hunk 6 above)
Comment (bhroam) [reply src/scheduler/buckets.c]: This one might deserve a comment.  The idea is if we can possibly run, we want to maintain that comment.  We don't want to return a comment for a placement set that the job can never run in.  That is the reason we were seeing never run messages.  There was a placement set that the job could run in, but just not right then.  We moved on and the final placement set we checked would never be possible.  We returned the comment for the final placement set.

**Comment by bhroam** [comment]
Looks good.  I sign off.

File: src/scheduler/buckets.c (hunk 5 above)
Comment (agrawalravi90) [reply src/scheduler/buckets.c]: Thanks for clarifying Bhroam

**Comment by agrawalravi90** [comment]
Thanks for adding the comment Arun, your changes look fine to me.

**Comment by arungrover** [comment]
@bhroam, @agrawalravi90 Thanks for the review, PR is ready for merge.

**Pull request merged**